db/python/tables/meta_table.py

```diff
@@ -0,0 +1,144 @@
+from io import BytesIO, StringIO
+from typing import Any, Callable
+
+# Unfortunately some of these libs have partially or completely missing
+# type annotations so mypy will have a few red underlines in this file :/
+import duckdb
+import pyarrow as pa
+import pyarrow.parquet as pq
+from databases.interfaces import Record
+
+from db.python.tables.base import DbBase
+from models.models import PRIMARY_EXTERNAL_ORG
+
+# Use this value to replace the PRIMARY_EXTERNAL_ORG value in the json object returned
+# from the query. DuckDB doesn't support column names which are an empty string.
+EXTERNAL_ORG_SENTINEL = '__PRIMARY__EXTERNAL__ORG__'
+
+
+class MetaTable(DbBase):
+    """
+    Generate a flat table with metadata for all entity tables
+    """
+
+    def external_id_query(self, table_alias: str):
+        """
+        Returns the query partial to aggregate external ids in to a json object
+        including replacing the primary external org with a sentinel value. This is
+        required because duckdb doesn't support column names which are an empty string.
+        """
+        return f"""
+            JSON_OBJECTAGG(
+                CASE
+                    WHEN {table_alias}.name = :primary_external_org
+                    THEN :primary_external_org_sentinel
+                    ELSE {table_alias}.name
+                END,
+                {table_alias}.external_id
+            ) AS external_ids
+        """
+
+    async def entity_meta_table(
+        self,
+        project: int,
+        query: str,
+        row_getter: Callable[[Record], dict[str, Any]],
+        has_external_ids: bool,
+        has_meta: bool,
+    ):
+        """
+        Return a flat tabular parquet file for the provied query. Optionally include
+        metadata and external_ids columns. This returns a BytesIO object with the
+        parquet file.
+        """
+
+        rows = await self.connection.fetch_all(
+            query,
+            {
+                'project': project,
+                'primary_external_org': PRIMARY_EXTERNAL_ORG,
```

Comment:
So am I right in understanding that the `EXTERNAL_ORG_SENTINEL` is a placeholder just to make duckdb happy?

Reply:
Yep, that's right. The default external id name is an empty string, and when duckdb reads the json rows, if we left it as an external string the column name would be an empty string, and sql columns can't be empty strings. Duckdb does kinda handle it by giving the column a name like `col1` (I can't remember exactly what it was) but the name is based on column position which isn't consistent.